src/uct/ib/mlx5/ib_mlx5_dv.c

```diff
@@ -22,6 +22,114 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
+                UCT_IB_DEVICE_FLAG_MELLANOX | UCT_IB_DEVICE_FLAG_MLX5_PRM |
+                UCT_IB_DEVICE_FLAG_DC_V1, 28, uct_ib_mlx5_device_open);
+UCT_DEVICE_SPEC(0x02c9, 4119, "ConnectX-5",
+                UCT_IB_DEVICE_FLAG_MELLANOX | UCT_IB_DEVICE_FLAG_MLX5_PRM |
+                UCT_IB_DEVICE_FLAG_DC_V2, 38, uct_ib_mlx5_device_open);
+UCT_DEVICE_SPEC(0x02c9, 4121, "ConnectX-5",
+                UCT_IB_DEVICE_FLAG_MELLANOX | UCT_IB_DEVICE_FLAG_MLX5_PRM |
+                UCT_IB_DEVICE_FLAG_DC_V2, 40, uct_ib_mlx5_device_open);
+UCT_DEVICE_SPEC(0x02c9, 4120, "ConnectX-5",
+                UCT_IB_DEVICE_FLAG_MELLANOX | UCT_IB_DEVICE_FLAG_MLX5_PRM |
+                UCT_IB_DEVICE_FLAG_DC_V2, 39, uct_ib_mlx5_device_open);
+UCT_DEVICE_SPEC(0x02c9, 41682, "ConnectX-5",
+                UCT_IB_DEVICE_FLAG_MELLANOX | UCT_IB_DEVICE_FLAG_MLX5_PRM |
+                UCT_IB_DEVICE_FLAG_DC_V2, 37, uct_ib_mlx5_device_open);
+_UCT_DEVICE_SPEC(0x02c9, 4122, "ConnectX-5",
+                UCT_IB_DEVICE_FLAG_MELLANOX | UCT_IB_DEVICE_FLAG_MLX5_PRM |
+                UCT_IB_DEVICE_FLAG_DC_V2, 36, uct_ib_mlx5_device_open, 1);
+
+ucs_status_t uct_ib_mlx5_device_open(struct ibv_context *ctx,
+                                     struct ibv_exp_device_attr *dev_attr,
+                                     uct_ib_device_spec_t *spec, int async_events
+                                     UCS_STATS_ARG(ucs_stats_node_t *stats_parent),
+                                     uct_ib_device_t **dev_p)
+{
+    struct ibv_qp_init_attr_ex qp_attr = {};
+    struct mlx5dv_qp_init_attr dv_attr = {};
+    uct_ib_mlx5_device_t *dev;
+    ucs_status_t status;
+    struct ibv_pd *pd;
+    struct ibv_cq *cq;
+    struct ibv_qp *qp;
+
+    dev = ucs_calloc(1, sizeof(*dev), "uct_ib_mlx5_device");
+    if (dev == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    *dev_p = &dev->super;
+    status = uct_ib_device_open(ctx, dev_attr, spec, async_events
+                                UCS_STATS_ARG(stats_parent), dev_p);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    pd = ibv_alloc_pd(ctx);
+    if (pd == NULL) {
+        ucs_error("ibv_alloc_pd() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err;
+    }
+
+    cq = ibv_create_cq(ctx, 1, NULL, NULL, 0);
+    if (cq == NULL) {
+        ucs_error("ibv_create_cq() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_cq;
+    }
+
+    qp_attr.send_cq              = cq;
+    qp_attr.recv_cq              = cq;
+    qp_attr.cap.max_send_wr      = 1;
+    qp_attr.cap.max_send_sge     = 1;
+    qp_attr.qp_type              = IBV_QPT_DRIVER;
+    qp_attr.comp_mask            = IBV_QP_INIT_ATTR_PD;
+    qp_attr.pd                   = pd;
+
+    dv_attr.comp_mask            = MLX5DV_QP_INIT_ATTR_MASK_DC;
+    dv_attr.dc_init_attr.dc_type = MLX5DV_DCTYPE_DCI;
+
+    qp = mlx5dv_create_qp(ctx, &qp_attr, &dv_attr);
+    if (qp) {
+        ibv_destroy_qp(qp);
+        dev->dc = 1;
```

Comment:
if it's just to remember DC support yes/no, maybe easier to add "flags" to uct_ib_device_t which could be filled by uct_ib_mlx5_device_open?
It can potentially contain also the flags from device_spec. e.g fill these flags in init time instead of statically.